### PR TITLE
[JUJU-3377] TrackedDB Enhancements

### PR DESCRIPTION
**Pull request by manadart**
This patch aims to address comment threads originally on https://github.com/juju/juju/pull/15331.

There are various mechanical changes:
- Use of constants.
- Relocation of logic closer to usage.
- Comment clarifications.

In addition to these, there is a material change to `TrackedDB` and its implementations.

We originally had a `DB` method that was used to execute a closure against the raw tracked `sql.DB`. This was something we didn't really want to do, but it was used by the lease expiry worker to circumvent the default retry strategy.

The problem with the method generally is that it opens us up to misuse such as eschewing context guards, that could potentially cause deadlocks.

Here, instead of `DB`, we use `TxnNoRetry`, which has the same signature as `Txn`, but (surprise!) will not use a retry strategy. This is not only tidier to use; we also retain both context handling and explicit transactions.

## QA steps

Bootstrap, enable HA and check that we've got controller unit leaders.

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

File: worker/leaseexpiry/worker.go [hunk 1]
```diff
@@ -100,48 +99,36 @@ func (w *expiryWorker) loop() error {
...
-				// all controller nodes. Retryable errors are those that occur due to
-				// locking or other contention. We know we will retry very soon,
-				// so just log and indicate success for these cases.
-				// Rethink this if the worker cardinality changes to be singular.
-				if txn.IsErrRetryable(err) {
-					w.logger.Debugf("ignoring error during lease expiry: %s", err.Error())
-					return nil
-				}
-				return errors.Trace(err)
+	err := w.trackedDB.TxnNoRetry(ctx, func(ctx context.Context, tx *sql.Tx) error {
```
Comment (manadart) [inline worker/leaseexpiry/worker.go:102]: The change can be seen here. the ergonomics of running without attempting retries is much nicer.

**Comment by hmlanigan** [comment]
@manadart I cherry-picked 5c9004 to be able to merge a PR into develop.

**Review by SimonRichardson — APPROVED** [review]

**Pull request merged**